lib/provisioning/Provisioner.js

```diff
@@ -0,0 +1,136 @@
+        {regex : /^[a-z\.0-9:-]+$/, example : 'example.com or localhost'},
+};
+
+Provisioner.prototype.validate = function(parameters) {
+    let parameterNames = ['matrix_room_id', 'remote_room_channel', 'remote_room_server'];
+    for (var i = 0; i < parameterNames.length; i++) {
+        let name = parameterNames[i];
+
+        let actual = parameters[name];
+        let valid = parameterValidation[name];
+
+        if (!actual) {
+            throw new Error(
+                `${name} not provided (like '${valid.example}').`
+            );
+        }
+
+        if (typeof actual !== 'string') {
+            throw new Error(
+                `${name} should be a string (like '${valid.example}').`
+            );
+        }
+
+        if (!actual.match(valid.regex)) {
+            throw new Error(
+                `Malformed ${name} ('${actual}'), should look like '${valid.example}'.`
+            );
+        }
+    }
+};
+
+// Link an IRC channel to a matrix room ID
+Provisioner.prototype.link = Promise.coroutine(function*(options) {
+    let ircDomain = options.remote_room_server;
+    let ircChannel = options.remote_room_channel;
+    let roomId = options.matrix_room_id;
+    let mappingLogId = `${roomId} <---> ${ircDomain}/${ircChannel}`;
+
+    log.info(`Provisioning link for room ${mappingLogId}`);
+
+    // Try to find the domain requested for linking
+    //TODO: ircDomain might include protocol, i.e. irc://irc.freenode.net
+    let server = this._ircBridge.getServer(ircDomain);
+
+    if (!server) {
+        throw new Error(`Server requested for linking not found ('${ircDomain}')`);
+    }
+
+    // Create rooms for the link
+    let ircRoom = new IrcRoom(server, ircChannel);
+    let mxRoom = new MatrixRoom(roomId);
+
+    let entry = yield this._ircBridge.getStore().getRoom(roomId, ircDomain, ircChannel);
+    if (!entry) {
+        this._ircBridge.getStore().storeRoom(ircRoom, mxRoom, 'provision');
+    }
+    else {
+        throw new Error(`Room mapping already exists (${mappingLogId},` +
+                        `origin = ${doc.data.origin})`);
+    }
+});
+
+// Unlink an IRC channel from a matrix room ID
+Provisioner.prototype.unlink = Promise.coroutine(function*(options) {
+    let ircDomain = options.remote_room_server;
+    let ircChannel = options.remote_room_channel;
+    let roomId = options.matrix_room_id;
+    let mappingLogId = `${roomId} <-/-> ${ircDomain}/${ircChannel}`;
+
+    log.info(`Provisioning unlink for room ${mappingLogId}`);
+
+    // Try to find the domain requested for unlinking
+    let server = this._ircBridge.getServer(ircDomain);
+
+    if (!server) {
+        throw new Error("Server requested for linking not found");
+    }
+
+    // Delete the room link
+    let entry = yield this._ircBridge.getStore()
+        .getRoom(roomId, ircDomain, ircChannel, 'provision');
+
+    if (entry) {
+        this._ircBridge.getStore().removeRoom(roomId, ircDomain, ircChannel, 'provision');
```

Comment:
You're dropping the `Promise` returned here to the floor. If `storeRoom` fails for whatever reason, we should fail the `link` Promise but we aren't.